sinabs/backend/dynapcnn/dynapcnn_network.py

```diff
@@ -1,108 +1,258 @@
+            prev_length = len(received_evts)
+            time.sleep(0.1)
+            received_evts.extend(self.samna_output_buffer.get_events())
+            if prev_length == len(received_evts):
+                break
+
+        # disable timestamp
+        disable_timestamps(self.device)
+
+        return received_evts
+
+    def forward(
+        self, x, return_complete: bool = False
+    ) -> Union[List["event"], Tensor, Dict[int, Dict[int, Tensor]]]:
+        """Forwards data through the `DynapcnnNetwork` instance.
+
+        If the network has been deployed on a Dynapcnn/Speck device the forward
+        pass happens on the devices. Otherwise the device will be simulated by
+        passing the data through the `DynapcnnLayer` instances.
+
+        Parameters
+        ----------
+        x: Tensor that serves as input to network. Is passed to all layers
+            that are marked as entry points
+        return_complete: bool that indicates whether all layer outputs should
+            be return or only those with no further destinations (default)
+
+        Returns
+        -------
+        The returned object depends on whether the network has been deployed
+        on chip. If this is the case, a flat list of samna events is returned,
+        in the order in which the events have been collected.
+        If the data is passed through the `DynapcnnLayer` instances, the output
+        depends on `return_complete` and on the network configuration:
+        * If `return_complete` is `True`, all layer outputs will be returned in a
+            dict, with layer indices as keys, and nested dicts as values, which
+            hold destination indices as keys and output tensors as values.
+        * If `return_complete` is `False` and there is only a single destination
+            in the whole network that is marked as final (i.e. destination
+            index in dynapcnn layer handler is negative), it will return the
+            output as a single tensor.
+        * If `return_complete` is `False` and no destination in the network
+            is marked as final, a warning will be raised and the function
+            returns an empty dict.
+        * In all other cases a dict will be returned that is of the same
+            structure as if `return_complete` is `True`, but only with entries
+            where the destination is marked as final.
+        """
+        if (
+            hasattr(self, "device")
+            and parse_device_id(self.device)[0] in ChipFactory.supported_devices
+        ):
+            return self.hw_forward(x)
         else:
-            self.dvs_input = False
+            # Forward pass through software DynapcnnLayer instance
+            return self.dynapcnn_module(x, return_complete=return_complete)
 
-        input_shape = infer_input_shape(layers, input_shape=input_shape)
-        assert len(input_shape) == 3, "infer_input_shape did not return 3-tuple"
+    def parameters(self) -> list:
```

Comment:
Why is this necessary? Inheriting from `nn.Module`, shouldn't there already be a `parameters` method?